### PR TITLE
feat: node resource topology scheduler plugin

### PR DESCRIPTION
**Pull request by WangZzzhe**
#### What type of PR is this?
Enhancements

#### What this PR does / why we need it:
add topologyMatch plugin to support topology-aware scheduling for both nativePolicy and dynamicPolicy
nativePolicy + single-numa-node, 
dynamicPolicy + dedicated_cores + numaBinding + single-numa-node,
dynamicPolicy + dedicated_cores + numaBinding + numaExclusive + single-numa-node,
dynamicPolicy + dedicated_cores + numaBinding + numeric,
dynamicPolicy + dedicated_cores + numaBinding + numaExclusive + numeric
are supported.

#### Which issue(s) this PR fixes:
https://github.com/kubewharf/katalyst-core/issues/160

### CODE REVIEW

**Comment by waynepeking348** [comment]
@caohe pls take your time to confirm whether we can merge this pr

**Comment by caohe** [comment]
> @caohe pls take your time to confirm whether we can merge this pr

Sure, we can merge this pr first https://github.com/kubewharf/enhanced-k8s/pull/12

Then https://github.com/kubewharf/katalyst-api/pull/39

**Comment by waynepeking348** [comment]
> > @caohe pls take your time to confirm whether we can merge this pr
> 
> Sure, we can merge this pr first [kubewharf/enhanced-k8s#12](https://github.com/kubewharf/enhanced-k8s/pull/12)
> 
> Then [kubewharf/katalyst-api#39](https://github.com/kubewharf/katalyst-api/pull/39)

enhanced-k8s has been merged, but the api still requires another approve, so you may need to approve that firstly @caohe

File: pkg/scheduler/plugins/noderesourcetopology/plugin.go [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+// should be registered for this plugin since a Pod update may free up resources
+// that make other Pods schedulable.
+func (tm *TopologyMatch) EventsToRegister() []framework.ClusterEvent {
+	// To register a custom event, follow the naming convention at:
+	// https://git.k8s.io/kubernetes/pkg/scheduler/eventhandlers.go#L403-L410
+	nrtGVK := fmt.Sprintf("customnoderesources.v1alpha1.%v", v1alpha1.GroupName)
+	return []framework.ClusterEvent{
+		{Resource: framework.Pod, ActionType: framework.Delete},
+		{Resource: framework.Node, ActionType: framework.Add | framework.UpdateNodeAllocatable},
+		{Resource: framework.GVK(nrtGVK), ActionType: framework.Add | framework.Update},
```
Comment (caohe) [inline pkg/scheduler/plugins/noderesourcetopology/plugin.go:128]: Is this CRD nrt or cnr?

Comment (WangZzzhe) [reply pkg/scheduler/plugins/noderesourcetopology/plugin.go]: > Is this CRD nrt or cnr?

cnr, it has been updated

File: pkg/scheduler/plugins/noderesourcetopology/reserve.go [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/klog/v2"
+	"k8s.io/kubernetes/pkg/scheduler/framework"
+
+	"github.com/kubewharf/katalyst-api/pkg/apis/node/v1alpha1"
+	"github.com/kubewharf/katalyst-core/pkg/scheduler/cache"
+	"github.com/kubewharf/katalyst-core/pkg/scheduler/util"
+)
+
+func (tm *TopologyMatch) Reserve(ctx context.Context, state *framework.CycleState, pod *corev1.Pod, nodeName string) *framework.Status {
```
Comment (caohe) [inline pkg/scheduler/plugins/noderesourcetopology/reserve.go:32]: How do we implement "pessimistic allocation" if multiple numa nodes are needed?

**Review by caohe — APPROVED** [review]

**Review by waynepeking348 — APPROVED** [review]

**Review by luomingmeng — APPROVED** [review]

**Review by csfldf — APPROVED** [review]

**Pull request merged**